### PR TITLE
feat(storage node): shard sync recovery path using BlobSyncHandler

### PR DESCRIPTION
**Pull request by halfprice**
- During shard sync, store missing blobs in a `pending_recover_blobs` table in shard.
- After shard sync, using blob recover to recover blobs in `pending_recover_blobs`.
- Non happy path currently do not preempt happy path currently.

Contributes to #259 #560 #705

### CODE REVIEW

File: crates/walrus-service/src/node/blob_sync.rs [hunk 1]
```diff
@@ -107,19 +126,31 @@ impl BlobSyncHandler {
...
             let cancel_token = CancellationToken::new();
             let synchronizer =
-                BlobSynchronizer::new(event, event_index, self.node.clone(), cancel_token.clone());
-
-            let sync_handle = tokio::spawn(
-                self.clone()
-                    .sync(synchronizer, start, self.semaphore.clone())
+                BlobSynchronizer::new(event.blob_id, self.node.clone(), cancel_token.clone());
+
+            let handler_clone = self.clone();
```
Comment (mlegner) [inline crates/walrus-service/src/node/blob_sync.rs]: Question: Is my understanding correct that we can easily clone the `BlobSyncHandler` here because it only contains `Arc`s? If that's the case, could we add a comment to the struct definition that this has to remain cheaply cloneable? (This seems to have been the case already before, so not related to your changes.)

File: crates/walrus-service/src/node/blob_sync.rs [hunk 2]
```diff
@@ -134,12 +165,64 @@ impl BlobSyncHandler {
...
+                    })
+                    .instrument(spawned_trace)
+                    .await;
+                handler_clone.blob_syncs_notify_reads.notify(&blob_id, &());
+                result
+            });
+            entry.insert(InProgressSyncHandle {
+                cancel_token,
+                blob_sync_handle: Some(sync_handle),
+            });
```
Comment (mlegner) [inline crates/walrus-service/src/node/blob_sync.rs]: Remark: This code is almost exactly the same as in the `start_sync` function; this could be deduplicated.

File: crates/walrus-service/src/node/blob_sync.rs (hunk 1 above)
Comment (halfprice) [reply crates/walrus-service/src/node/blob_sync.rs]: Yes, that's correct. Added a struct level doc.

File: crates/walrus-service/src/node/blob_sync.rs (hunk 2 above)
Comment (halfprice) [reply crates/walrus-service/src/node/blob_sync.rs]: Extracted it out

File: crates/walrus-service/src/test_utils.rs [hunk 3]
```diff
@@ -714,6 +714,19 @@ impl TestClusterBuilder {
         self
     }
 
+    /// Sets the individual event providers for each storage node.
```
Comment (mlegner) [inline crates/walrus-service/src/test_utils.rs:718]: Nit: Can we document that the length of the slice should be the same as the number of storage nodes?

Also, in many cases the caller could move ownership of the providers such that we wouldn't have to clone them. But as this is a test function, that probably doesn't matter too much.

File: crates/walrus-service/src/node/storage/shard.rs [hunk 4]
```diff
@@ -140,6 +154,16 @@ impl ShardStorage {
...
+        let pending_recover_slivers_cf_name = pending_recover_slivers_column_family_name(id);
+        if database
+            .cf_handle(&pending_recover_slivers_cf_name)
+            .is_none()
+        {
+            let (_, options) = Self::pending_recover_slivers_column_family_options(id, db_config);
+            database
+                .create_cf(&pending_recover_slivers_cf_name, &options)
+                .map_err(typed_store_err_from_rocks_err)?;
+        }
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs:177]: Remark: I feel we have more and more code duplication in this function. Maybe at some point (not this PR), we can deduplicate these operations through a macro or so.

File: crates/walrus-service/src/node/storage/shard.rs [hunk 5]
```diff
@@ -387,29 +429,38 @@ impl ShardStorage {
...
-                    sliver_type == SliverType::Secondary,
-                )
-            }
-            None => (None, false),
-        };
+        let (mut last_synced_blob_id, skip_primary, skip_secondary) =
+            if shard_status == ShardStatus::ActiveRecover {
+                // We are in recovery mode. Skip happy path syncing and directly jump to recover
+                // missing blobs.
+                (None, true, true)
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs]: Suggestion: Replace this triple by an enum to make the current status more explicit:
```rust
enum ShardSyncStatus {
    Recovery,
    Primary {
        last_synced_blob_id: Option<BlobId>,
    },
    Secondary {
        last_synced_blob_id: Option<BlobId>,
    },
}
```
Then you can do matches like 
```rust
if let ShardSyncStatus::Primary{last_synced_blob_id}) = shard_sync_status { ... }
```

File: crates/walrus-service/src/node/storage/shard.rs [hunk 6]
```diff
@@ -421,13 +472,44 @@ impl ShardStorage {
...
+                    "sync shard simulated sync failure after start recovery"
+                )))
+            });
+
+            loop {
+                self.recover_missing_blobs(blob_sync_handler.clone())
+                    .await?;
+
+                if self.pending_recover_slivers.is_empty() {
+                    // TODO: check recovered blob consistency.
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs]: Question: Isn't this handled by the blob sync internally?

File: crates/walrus-service/src/node/storage/shard.rs [hunk 7]
```diff
@@ -421,13 +472,44 @@ impl ShardStorage {
...
+                    break;
+                }
+                tracing::warn!(
+                    "Recovering missing blobs for shard: {} still misses blobs.
+                    Retrying in 60 seconds.",
+                    self.id
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs]: Nit: The shard ID is already included in the span and could be removed here.

File: crates/walrus-service/src/node/storage.rs [hunk 8]
```diff
@@ -178,6 +178,7 @@ pub struct DatabaseConfig {
     shard: DatabaseTableOptions,
     shard_status: DatabaseTableOptions,
     shard_sync_progress: DatabaseTableOptions,
+    pending_recover_slivers: DatabaseTableOptions,
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage.rs:180]: Remark: We now have so many tables, it becomes very tedious to change the config as this has to be done for every table separately. It would be nice if in the config file one could specify custom "default" and "blob-optimized" options, which would apply to all corresponding tables that don't have a more specific config.

File: crates/walrus-service/src/node/storage.rs [hunk 9]
```diff
@@ -224,6 +231,57 @@ pub struct Storage {
...
+            iter,
+            before_epoch,
+            only_certified,
+        }
+    }
+}
+
+impl Debug for BlobInfoIter<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("BlobInfoIter").finish()
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage.rs]: Suggestion: Include the `before_epoch` and `only_certified` in the debug output.

File: crates/walrus-service/src/node/storage.rs [hunk 10]
```diff
@@ -224,6 +231,57 @@ pub struct Storage {
...
+        let mut item = self.iter.next();
+        if !self.only_certified {
+            return item;
+        }
+
+        while let Some(Ok((_, ref blob_info))) = item {
+            if blob_info.is_certified()
+                && blob_info
+                    .status_changing_epoch(blob_info::BlobCertificationStatus::Certified)
+                    .unwrap()
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage.rs]: Suggestion: Can you replace this with an `expect` describing why this can never panic?

File: crates/walrus-service/src/node/storage.rs [hunk 11]
```diff
@@ -358,14 +416,21 @@ impl Storage {
...
-        match range {
-            Some(range) => self.blob_info.safe_range_iter(range),
-            None => self.blob_info.safe_iter(),
-        }
+        before_epoch: Epoch,
+        last_synced_blob_id: Option<BlobId>,
+    ) -> BlobInfoIter {
+        BlobInfoIter::new(
+            Box::new(match last_synced_blob_id {
+                Some(starting_blob_id) => self
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage.rs:428]: Nit: Can we use the same name `last_synced_blob_id` here?

File: crates/walrus-service/src/node/storage.rs [hunk 12]
```diff
@@ -224,6 +231,57 @@ pub struct Storage {
     config: DatabaseConfig,
 }
 
+/// A iterator over the blob info table.
+pub struct BlobInfoIter<'a> {
+    iter: Box<dyn Iterator<Item = Result<(BlobId, BlobInfo), TypedStoreError>> + Send + 'a>,
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage.rs:235]: Question: Do we actually need the trait object here? AFAICT, we always have the same iterator type.

File: crates/walrus-service/src/node/storage.rs [hunk 13]
```diff
@@ -358,14 +416,21 @@ impl Storage {
...
-        }
+        before_epoch: Epoch,
+        last_synced_blob_id: Option<BlobId>,
+    ) -> BlobInfoIter {
+        BlobInfoIter::new(
+            Box::new(match last_synced_blob_id {
+                Some(starting_blob_id) => self
+                    .blob_info
+                    .safe_range_iter((Excluded(starting_blob_id), Unbounded)),
+                None => self.blob_info.safe_iter(),
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage.rs:431]: Question: What happens to these iterators when data is written to the DB?

File: crates/walrus-service/src/node/storage.rs [hunk 14]
```diff
@@ -1245,4 +1310,96 @@ pub(crate) mod tests {
...
+        let new_epoch = 3;
+
+        let blob_ids = [
+            BlobId([1; 32]), // Not certified.
+            BlobId([2; 32]), // Not exist.
+            BlobId([3; 32]), // Certified within epoch 2
+            BlobId([4; 32]), // Certified after epoch 2
+            BlobId([5; 32]), // Invalid
+            BlobId([6; 32]), // Certified within epoch 2
+            BlobId([7; 32]), // Not exist.
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage.rs]: Nit: Can we shift the values by 1 so that `blob_ids[x] == BlobId([x;32])`? This would help with following the code.

File: crates/walrus-service/src/node/storage.rs [hunk 15]
```diff
@@ -1245,4 +1310,96 @@ pub(crate) mod tests {
...
+            (blob_ids[3], certified_blob_info(3)),
+            (blob_ids[4], invalid_blob_info(2)),
+            (blob_ids[5], certified_blob_info(2)),
+        ]);
+
+        let mut batch = blob_info.batch();
+        batch.insert_batch(&blob_info, blob_info_map.iter())?;
+        batch.write()?;
+
+        for blob_id in blob_ids.iter().take(2) {
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage.rs:1394]: Nit: Can we also test the `None` case?

File: crates/walrus-service/src/node/storage/shard.rs [hunk 16]
```diff
@@ -6,20 +6,28 @@
...
 use anyhow::Context;
 use fastcrypto::traits::KeyPair;
+use futures::{stream::FuturesUnordered, StreamExt};
 use regex::Regex;
 use rocksdb::{Options, DB};
 use serde::{Deserialize, Serialize};
 use typed_store::{
-    rocks::{errors::typed_store_err_from_rocks_err, DBBatch, DBMap, ReadWriteOptions, RocksDB},
+    rocks::{
+        be_fix_int_ser as ToRocksDbKey,
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs]: Nit: Seems a bit weird to me to rename this looking like a type. Can we use `to_rocks_db_key` instead?

File: crates/walrus-service/src/node/storage/shard.rs [hunk 17]
```diff
@@ -456,10 +538,14 @@ impl ShardStorage {
...
-                .context("Scanning certified blobs encountered error")?
-        {
+
+        let mut blob_info_iter = node
+            .storage
+            .certified_blob_info_iter_before_epoch(epoch, last_synced_blob_id);
+
+        let mut next_blob_info = blob_info_iter.next().transpose()?;
+        while next_blob_info.is_some() {
+            let next_starting_blob_id = next_blob_info.as_ref().unwrap().0;
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs]: Nit: Instead of the `is_some` + `unwrap` construct, we could use a `while let`:
```suggestion
        while let Some((next_starting_blob_id, _)) = &next_blob_info {
```

File: crates/walrus-service/src/node/storage/shard.rs [hunk 18]
```diff
@@ -456,10 +538,14 @@ impl ShardStorage {
...
-            next_certified_blob_id_before_epoch(&node.storage, epoch, last_synced_blob_id)
-                .context("Scanning certified blobs encountered error")?
-        {
+
+        let mut blob_info_iter = node
+            .storage
+            .certified_blob_info_iter_before_epoch(epoch, last_synced_blob_id);
+
+        let mut next_blob_info = blob_info_iter.next().transpose()?;
+        while next_blob_info.is_some() {
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs]: Important: This loop now has almost 100 lines. I think it is time to refactor this a bit to keep it readable.

File: crates/walrus-service/src/node/storage/shard.rs [hunk 19]
```diff
@@ -482,8 +568,7 @@ impl ShardStorage {
                 )
                 .await?;
             for blob in fetched_slivers.iter() {
-                tracing::debug!("Synced blob id: {} to before epoch: {}.", blob.0, epoch,);
-                //TODO(#705): Track missing blobs.
+                tracing::info!("Synced blob id: {} to before epoch: {}.", blob.0, epoch,);
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs]: Question: What is the motivation of increasing this to info level? There will be a lot of these messages...

File: crates/walrus-service/src/node/storage/shard.rs [hunk 20]
```diff
@@ -536,13 +629,129 @@ impl ShardStorage {
...
         }
 
+        if next_blob_info.is_none() {
+            return Ok(());
+        }
+
+        let mut batch = match sliver_type {
+            SliverType::Primary => self.primary_slivers.batch(),
+            SliverType::Secondary => self.secondary_slivers.batch(),
+        };
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs]: Question: Why do you create the batch from these tables?

File: crates/walrus-service/src/node/storage/shard.rs [hunk 21]
```diff
@@ -536,13 +629,129 @@ impl ShardStorage {
...
+        batch.write()?;
+
+        Ok(())
+    }
+
+    // Given a iterator over blob info table that is currently pointing to `next_blob_info`,
+    // and `fetched_blob_id` returned from the remote storage node, checks if there are any
+    // missing blobs between `next_blob_info` and `fetched_blob_id`.
+    // Returns the next blob info that is greater than `fetched_blob_id` or None if there are no.
+    fn check_missing_blobs(
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs]: Nit: Can we rename this a bit? AFAIU, the purpose of this function is to add the missing blobs to the DB to be recovered later.

File: crates/walrus-service/src/node/storage/shard.rs [hunk 22]
```diff
@@ -536,13 +629,129 @@ impl ShardStorage {
...
+        &self,
+        blob_info_iter: &mut BlobInfoIter,
+        mut next_blob_info: Option<(BlobId, BlobInfo)>,
+        fetched_blob_id: BlobId,
+        sliver_type: SliverType,
+        db_batch: &mut DBBatch,
+    ) -> Result<Option<(BlobId, BlobInfo)>, TypedStoreError> {
+        let Some((mut next_blob_id, _)) = next_blob_info else {
+            assert!(blob_info_iter.next().transpose()?.is_none());
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs]: Important: Do we really want to panic here? Can we change this to either an error or a `debug_assert`? Also, maybe we could just change the type of that parameter to `(BlobId, BlobInfo)>`.

File: crates/walrus-service/src/node/storage/shard.rs [hunk 23]
```diff
@@ -536,13 +629,129 @@ impl ShardStorage {
...
+        }
+        batch.write()?;
+
+        Ok(())
+    }
+
+    // Given a iterator over blob info table that is currently pointing to `next_blob_info`,
+    // and `fetched_blob_id` returned from the remote storage node, checks if there are any
+    // missing blobs between `next_blob_info` and `fetched_blob_id`.
+    // Returns the next blob info that is greater than `fetched_blob_id` or None if there are no.
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs]: Nit: Can we make these docstrings instead of normal comments? (Also applies to other functions.)

File: crates/walrus-service/src/node/storage/shard.rs [hunk 24]
```diff
@@ -536,13 +629,129 @@ impl ShardStorage {
...
+            "Start recovering missing blob {} in shard {}",
+            blob_id,
+            self.id
+        );
+        // Note that `blob_sync_handler` tries to recover ALL missing shards for the given blob
+        // id. `start_sync_for_blob_recovery` can be called multiple times, but there will
+        // be only one active recovery task for the given blob id.
+        blob_sync_handler
+            .start_sync_for_blob_recovery(blob_id, start_time)
+            .await;
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs:776]: Important: Essentially this means that for each blob missing after shard sync, we go through all shard storages and check if the corresponding slivers are stored. This seems a massive waste of resources if we just want to recover one sliver pair, especially for large storage nodes that can have tens of shards.

File: crates/walrus-service/src/node/storage/shard.rs [hunk 25]
```diff
@@ -536,13 +629,129 @@ impl ShardStorage {
...
+                    (SliverType::Primary, blob_id),
+                    (SliverType::Secondary, blob_id),
+                ],
+            )?;
+            batch.write()?;
+        }
+
+        // If sliver is not recovered, it won't be removed from the `pending_recover_slivers` table.
+        // Later retry will revisit this blob again.
         Ok(())
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs:791]: Important: AFAIU, the blob sync spawns a new task, so we will not have the sliver stored after starting the blob sync. OTOH, we probably don't have to implement retries here as the blob sync itself takes care of that.

File: crates/walrus-service/src/node/storage/shard.rs [hunk 26]
```diff
@@ -989,110 +1199,97 @@ mod tests {
...
     }
 
-    #[tokio::test]
-    async fn test_next_certified_blob_id_before_epoch() -> TestResult {
+    pub fn random_blob_id() -> BlobId {
+        let mut bytes = [0; 32];
+        rand::thread_rng().fill_bytes(&mut bytes);
+        BlobId(bytes)
+    }
```
Comment (mlegner) [inline crates/walrus-service/src/node/storage/shard.rs]: Hint: There already exists such a function: `walrus_core::test_utils::random_blob_id()`.

File: crates/walrus-service/src/node/storage.rs (hunk 12 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage.rs:235]: Yes. The issue is that DBMap::SafeIter is a private type.

File: crates/walrus-service/src/node/storage.rs (hunk 9 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage.rs]: Done

File: crates/walrus-service/src/node/storage.rs (hunk 10 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage.rs]: Done

File: crates/walrus-service/src/node/storage.rs (hunk 11 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage.rs:428]: Done

File: crates/walrus-service/src/node/storage.rs (hunk 13 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage.rs:431]: Iterator operates on a snapshot of a DB. So if there are new writes after the iterator is created, they won't be observed.

File: crates/walrus-service/src/node/storage.rs (hunk 14 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage.rs]: Done

File: crates/walrus-service/src/node/storage.rs (hunk 15 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage.rs:1394]: Done

File: crates/walrus-service/src/node/storage/shard.rs (hunk 16 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs]: Done

File: crates/walrus-service/src/node/storage/shard.rs (hunk 4 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs:177]: Added a TODO

File: crates/walrus-service/src/node/storage/shard.rs (hunk 6 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs]: This is not blob consistency, but that the synced blob is consistent with what the shard wants to sync: it has received all blobs. 

This meant as an invariant check only in tests.

Revised the TODO.

File: crates/walrus-service/src/node/storage/shard.rs (hunk 7 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs]: Done

File: crates/walrus-service/src/node/storage/shard.rs (hunk 19 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs]: Reverted. Was used in debugging...

File: crates/walrus-service/src/node/storage/shard.rs (hunk 20 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs]: Because I blindly accepted a code suggestion... Changed.

File: crates/walrus-service/src/node/storage/shard.rs (hunk 23 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs]: Done

File: crates/walrus-service/src/node/storage/shard.rs (hunk 22 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs]: I think so because next_blob_info should be the last item read from blob_info_iter, and it this is not none, it's a invariant violation. This check is meant for catching implementation bugs. Changed to debug_assert and added more comments.

File: crates/walrus-service/src/node/storage/shard.rs (hunk 25 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs:791]: This is not a TODO, but a comment.

`recover_missing_blobs` waits until all blobs are synced. If there is any missing blobs, it'll restart recovering again. The comment here is for that case.

File: crates/walrus-service/src/test_utils.rs (hunk 3 above)
Comment (halfprice) [reply crates/walrus-service/src/test_utils.rs:718]: Done

File: crates/walrus-service/src/node/storage/shard.rs (hunk 26 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs]: Yeah, I should have pointed it out. The issue with that one is that it always returns the same blob id since it always uses the same seed. I moved `random()` into BlobID, so that it's easier to find when next time people need it. Maybe we can replace all the random_blob_id() with this one.

File: crates/walrus-service/src/node/storage/shard.rs (hunk 21 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs]: update it to `check_and_record_missing_blobs`

File: crates/walrus-service/src/node/storage/shard.rs (hunk 18 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs]: Done. Break it into several helper functions.

File: crates/walrus-service/src/node/storage/shard.rs (hunk 5 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs]: Done

File: crates/walrus-service/src/node/storage/shard.rs (hunk 6 above)
Comment (mlegner) [reply crates/walrus-service/src/node/storage/shard.rs]: Ah, I see, thanks for the clarification.

File: crates/walrus-service/src/node/storage.rs (hunk 8 above)
Comment (mlegner) [reply crates/walrus-service/src/node/storage.rs:180]: Clarification: This is not something for this PR but for a follow-up.

File: crates/walrus-service/src/node/storage.rs (hunk 12 above)
Comment (mlegner) [reply crates/walrus-service/src/node/storage.rs:235]: Hmm... But couldn't we make the `BlobInfoIter` generic over an inner iterator implementing `Iterator<Item = Result<(BlobId, BlobInfo), TypedStoreError>`? Something like this:

```rust
/// A iterator over the blob info table.
pub struct BlobInfoIter<I> {
    iter: I,
    before_epoch: Epoch,
    only_certified: bool,
}

impl<I> BlobInfoIter<I>
where
    I: Iterator<Item = Result<(BlobId, BlobInfo), TypedStoreError>>,
{
    pub fn new(iter: I, before_epoch: Epoch, only_certified: bool) -> Self {
        Self {
            iter,
            before_epoch,
            only_certified,
        }
    }
}
```

File: crates/walrus-service/src/node/storage.rs (hunk 13 above)
Comment (mlegner) [reply crates/walrus-service/src/node/storage.rs:431]: OK, thanks for the insight! Just for the future: We need to make sure that the shard sync and blob sync still work properly if blobs are deleted.

File: crates/walrus-service/src/node/storage/shard.rs (hunk 26 above)
Comment (mlegner) [reply crates/walrus-service/src/node/storage/shard.rs]: Ah, good point. Yes, maybe you could just replace the body of the `random_blob_id` with this one?

File: crates/walrus-service/src/node/storage/shard.rs (hunk 25 above)
Comment (mlegner) [reply crates/walrus-service/src/node/storage/shard.rs:791]: Hmm... And as the `start_sync_for_blob_recovery` doesn't actually start multiple tasks for the same blob, we just call it again when we check later?

File: crates/walrus-service/src/node/blob_sync.rs [hunk 27]
```diff
@@ -134,12 +148,76 @@ impl BlobSyncHandler {
         Ok(())
     }
 
+    /// An entry point for blob sync initiated by the shard sync process.
+    #[tracing::instrument(skip_all, fields(otel.kind = "PRODUCER"))]
+    pub fn start_sync_for_blob_recovery(
+        &self,
+        blob_id: BlobId,
+        start: tokio::time::Instant,
+    ) -> Registration<BlobId, ()> {
```
Comment (mlegner) [inline crates/walrus-service/src/node/blob_sync.rs:157]: Question: What is the purpose of the return value? We actually immediately drop it in the currently only place where we call this function.

Comment (halfprice) [reply crates/walrus-service/src/node/blob_sync.rs:157]: This is for the caller to be able wait on the task associated with the BlobID. It was await at the caller so it's not dropped immediately.

The new version of the code doesn't involve this.

File: crates/walrus-service/src/node/storage/shard.rs (hunk 25 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs:791]: No longer relevent since we are moving with #746

File: crates/walrus-service/src/node/storage/shard.rs (hunk 26 above)
Comment (halfprice) [reply crates/walrus-service/src/node/storage/shard.rs]: Done in the #746
